{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 11\n",
    "###Â Useful Functions and Methods"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**1. During your research in ion channels you stumble upon the SwissProt entry (CIQ3_HUMAN) in appendix 1. You notice in the feature table (FT) that several variants and mutations of this gene exists. You want to take a closer look at this and decide that the first step is to extract the original amino acid sequence and all variations (full sequence with the appropiate amino acid changed) hereoff and put the result in a fasta file. Do not consider combinations. Since you probably are going to do this on a lot of SwissProt entries, you decide to make a program in python.** \n",
    "\n",
    "    a) Describe which keywords/patterns you will be looking for when parsing the file searching for the variants/mutations and the sequence.\n",
    "    b) Describe a method, perhaps using pseudo code, to extract the sequence and variations/mutants and print all the different full sequences in fasta. \n",
    "    c) Implement your method in python.\n",
    "    d) What kind of error checking could/should you include in your program ? Here you should name every check, which is relevant to the task, not every check it is possible to make.\n",
    "    e) In what way could you generalize or extend the program ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Printing the results in an output file (aa_seq.txt)...\n"
     ]
    }
   ],
   "source": [
    "infile = open(\"/Users/laurasansc/github/learning_julia/files/appendix1.txt\", \"r\")\n",
    "\n",
    "seq = \"\"\n",
    "flag = false # Define a variable for the sequence\n",
    "mutation = Dict()\n",
    "id_prot = \"\"\n",
    "\n",
    "for line in eachline(infile)\n",
    "    if occursin(\"ID\", line)\n",
    "        id_line = split(line)\n",
    "        id_prot = id_line[2]\n",
    "\n",
    "    elseif occursin(\"FT\", line)\n",
    "        ft_line = split(line)\n",
    "        if ft_line[2] == \"VARIANT\" || ft_line[2] == \"MUTAGEN\"\n",
    "            result_match = match(r\"^FT\\s+(\\w+)\\s+\\d+\\s+(\\d+)\\s+(\\w+)\\s*?[-][>]\\s*?(\\w+)\", line)\n",
    "            \n",
    "            if ft_line[3] == ft_line[4]\n",
    "                mutation[ft_line[3]] = (result_match[3], result_match[4])\n",
    "            end\n",
    "        end\n",
    "    \n",
    "    elseif occursin(\"//\", line)\n",
    "            flag = false\n",
    "    elseif flag == true\n",
    "            line = strip(line)\n",
    "            line = replace(line, \" \" => \"\")\n",
    "            seq = string(seq, line)\n",
    "    elseif occursin(\"SQ\", line)\n",
    "            flag = true\n",
    "    end \n",
    "end\n",
    "\n",
    "println(\"Printing the results in an output file (aa_seq.txt)...\")\n",
    "\n",
    "outfile = open(\"aa_seq.txt\", \"w\")\n",
    "\n",
    "write(outfile,\">\",id_prot,\"\\tWILD TYPE\\n\")\n",
    "\n",
    "for i in 1:60:length(seq)\n",
    "    if (i+60) > length(seq)\n",
    "        write(outfile,seq[i:length(seq)], \"\\n\")\n",
    "    else\n",
    "        write(outfile, seq[i:i+60], \"\\n\")\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "for (key, value) in mutation\n",
    "    value = collect(value)\n",
    "    k = parse.(Int64, key)\n",
    "\n",
    "    write(outfile, \">\",id_prot,\" MUTATION position: \", key, \" (\", value[1],\" => \", value[2], \")\\n\")\n",
    "    mut_sequence = collect(seq)\n",
    "    splice!(mut_sequence, k, value[1])\n",
    "    mut_sequence = join(mut_sequence, \"\")\n",
    "    for i in 1:60:length(mut_sequence)\n",
    "        if (i+60) > length(seq)\n",
    "            write(outfile,mut_sequence[i:length(seq)], \"\\n\")\n",
    "        else\n",
    "            write(outfile, mut_sequence[i:i+60], \"\\n\")\n",
    "        end\n",
    "    end\n",
    "end\n",
    "close(infile)\n",
    "close(outfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**2. You have earlier in your career made a splendid program, that calculate various scores based on amino acid sequence features. The output from your program can be seen in appendix 3, and consist of an accession number followed by 6 numbers between 0 and 1 per line (tab separated). You want to find the accession numbers with the highest and lowest average scores (average of the 6 numbers). However, you want to exclude any genes on your negative list from your calculations. These genes are listed as SwissProt IDs in appendix 4. Since GenBank accession numbers and SwissProt IDs are not identical, you need to translate between them in order to solve your problem. Fortunately you have a file, that does just that, see appendix 5, where the first item on the line is a SwissProt ID, second item is irrelevant, and third is the corresponding GenBank accession number.**\n",
    "\n",
    "    a) Describe a method, perhaps using pseudo code, to find the data.\n",
    "    b) Implement your method in python.\n",
    "    c) Have you made any assumptions about the data in your algoritm ? Which ? Why ? Are they reasonable assumptions (explain) ? Could/should you do away with them (by changing the code) ?\n",
    "    d) Usually, when you have this kind of problem, you want the highest 10 and lowest 10 average scores, not just the top and bottom average score. How would you solve this problem ? Will it change any assumptions i c) ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "infile = open(\"/Users/laurasansc/github/learning_julia/files/appendix3.txt\", \"r\")\n",
    "translation_ID = open(\"/Users/laurasansc/github/learning_julia/files/appendix5.txt\", \"r\")\n",
    "negative = open(\"/Users/laurasansc/github/learning_julia/files/appendix4.txt\", \"r\")\n",
    "\n",
    "# get average of app3\n",
    "# exclude from app3 the ones in app4\n",
    "# use app5 to translate from 1 id to the other\n",
    "\n",
    "#exclusion list\n",
    "exclusion_list = [line for line in eachline(negative)]\n",
    "\n",
    "# dict to translate from SwissProtID to GenBankID and viceversa\n",
    "translate = Dict()\n",
    "for line in eachline(translation_ID)\n",
    "\tentry = split(line)\n",
    "\ttranslate[entry[3]] = entry[1]\n",
    "end\n",
    "\n",
    "\n",
    "# dict of the scores we want to calculate (raw data, without exclusions)\n",
    "scores = Dict()\n",
    "for line in eachline(infile)\n",
    "    entry = split(line)\n",
    "    scores[entry[1]] = [] \n",
    "    for i in 2:length(entry)\n",
    "        push!(scores[entry[1]], parse.(Float64, entry[i]))\n",
    "    end\n",
    "end\n",
    "\n",
    "# calculate average make dict with genID and swisprotID and then exclude negatives\n",
    "average = Dict()\n",
    "for (k,v) in scores\n",
    "    if haskey(translate, k) == true\n",
    "        if (translate[k] in exclusion_list) == false     \n",
    "            average[(k,translate[k])] = sum(v)/length(v)\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "#open outfile\n",
    "outfile = open(\"average_scores.txt\", \"w\")\n",
    "\n",
    "\n",
    "# sort and get top 3 and the bottom 3\n",
    "sorted_scores = sort(collect(values(average)),rev=true)\n",
    "\n",
    "# X = how many top/bottom scores to show\n",
    "x = 3\n",
    "# get top and bottom scores\n",
    "top_scores = sorted_scores[1:x]\n",
    "bottom_scores = sorted_scores[end-x:end]\n",
    "\n",
    "#write outfile\n",
    "write(outfile, \"TOP SCORES \\n\")\n",
    "for i in 1:length(top_scores)\n",
    "    for (k,v) in average\n",
    "        key = collect(k)\n",
    "        if v == top_scores[i]\n",
    "            v = round(v,digits = 4)\n",
    "            line = string(key[1],\"\\t\",key[2],\"\\t\",v,\"\\n\")\n",
    "            write(outfile, line)\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "write(outfile, \"BOTTOM SCORES \\n\")\n",
    "for i in 1:length(bottom_scores)\n",
    "    for (k,v) in average\n",
    "        key = collect(k)\n",
    "        if v == bottom_scores[i]\n",
    "            v = round(v,digits = 4)\n",
    "            line = string(key[1],\"\\t\",key[2],\"\\t\",v,\"\\n\")\n",
    "            write(outfile, line)\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "close(translation_ID)\n",
    "close(negative)\n",
    "close(infile)\n",
    "close(outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.5.0",
   "language": "julia",
   "name": "julia-1.5"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.5.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
