{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercises 7\n",
    "### Sets and Dictionaries"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**1.Create a dictionary where the keys are codons and the value are the one-letter-code for the amino acids. The dictionary will function as a look-up table. You can find a codon list here.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter a codon and I will send the amin acid. Will return X if there is no aminoacid for the codon. \n",
      "stdin> f+\n",
      "You entered a wrong codon\n"
     ]
    }
   ],
   "source": [
    "codon_dict = Dict([\n",
    "    (\"ATA\",\"I\"), (\"ATC\",\"I\"), (\"ATT\",\"I\"), (\"ATG\",\"M\"),\n",
    "    (\"ACA\",\"T\"), (\"ACC\",\"T\"), (\"ACG\",\"T\"), (\"ACT\",\"T\"),\n",
    "    (\"AAC\",\"N\"), (\"AAT\",\"N\"), (\"AAA\",\"K\"), (\"AAG\",\"K\"),\n",
    "    (\"AGC\",\"S\"), (\"AGT\",\"S\"), (\"AGA\",\"R\"), (\"AGG\",\"R\"),\n",
    "    (\"CTA\",\"L\"), (\"CTC\",\"L\"), (\"CTG\",\"L\"), (\"CTT\",\"L\"),\n",
    "    (\"CCA\",\"P\"), (\"CCC\",\"P\"), (\"CCG\",\"P\"), (\"CCT\",\"P\"),\n",
    "    (\"CAC\",\"H\"), (\"CAT\",\"H\"), (\"CAA\",\"Q\"), (\"CAG\",\"Q\"),\n",
    "    (\"CGA\",\"R\"), (\"CGC\",\"R\"), (\"CGG\",\"R\"), (\"CGT\",\"R\"),\n",
    "    (\"GTA\",\"V\"), (\"GTC\",\"V\"), (\"GTG\",\"V\"), (\"GTT\",\"V\"),\n",
    "    (\"GCA\",\"A\"), (\"GCC\",\"A\"), (\"GCG\",\"A\"), (\"GCT\",\"A\"),\n",
    "    (\"GAC\",\"D\"), (\"GAT\",\"D\"), (\"GAA\",\"E\"), (\"GAG\",\"E\"),\n",
    "    (\"GGA\",\"G\"), (\"GGC\",\"G\"), (\"GGG\",\"G\"), (\"GGT\",\"G\"),\n",
    "    (\"TCA\",\"S\"), (\"TCC\",\"S\"), (\"TCG\",\"S\"), (\"TCT\",\"S\"),\n",
    "    (\"TTC\",\"F\"), (\"TTT\",\"F\"), (\"TTA\",\"L\"), (\"TTG\",\"L\"),\n",
    "    (\"TAC\",\"Y\"), (\"TAT\",\"Y\"), (\"TAA\",\"STOP\"), (\"TAG\",\"STOP\"),\n",
    "    (\"TGC\",\"C\"), (\"TGT\",\"C\"), (\"TGA\",\"STOP\"), (\"TGG\",\"W\")])\n",
    "\n",
    "println(\"Enter a codon and I will send the amin acid. Will return X if there is no aminoacid for the codon. \")\n",
    "query = uppercase(readline())\n",
    "\n",
    "if length(query) != 3\n",
    "    println(\"You entered a wrong codon\")\n",
    "    #exit(code = 0)\n",
    "else\n",
    "    println(get(codon_dict, query, \"X\"))\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**2. Use the dictionary from the previous exercise in a program, that translates all the nucleotide fasta entries in dna7.fsa to amino acid sequence. Save the results in a file aa7.fsa in fasta format. Remember to keep the 'headlines' for each entry and add 'Amino Acid Sequence' to each of them. The STOP codon is NOT a part of the amino acid sequence.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_dict = Dict([\n",
    "    (\"ATA\",\"I\"), (\"ATC\",\"I\"), (\"ATT\",\"I\"), (\"ATG\",\"M\"),\n",
    "    (\"ACA\",\"T\"), (\"ACC\",\"T\"), (\"ACG\",\"T\"), (\"ACT\",\"T\"),\n",
    "    (\"AAC\",\"N\"), (\"AAT\",\"N\"), (\"AAA\",\"K\"), (\"AAG\",\"K\"),\n",
    "    (\"AGC\",\"S\"), (\"AGT\",\"S\"), (\"AGA\",\"R\"), (\"AGG\",\"R\"),\n",
    "    (\"CTA\",\"L\"), (\"CTC\",\"L\"), (\"CTG\",\"L\"), (\"CTT\",\"L\"),\n",
    "    (\"CCA\",\"P\"), (\"CCC\",\"P\"), (\"CCG\",\"P\"), (\"CCT\",\"P\"),\n",
    "    (\"CAC\",\"H\"), (\"CAT\",\"H\"), (\"CAA\",\"Q\"), (\"CAG\",\"Q\"),\n",
    "    (\"CGA\",\"R\"), (\"CGC\",\"R\"), (\"CGG\",\"R\"), (\"CGT\",\"R\"),\n",
    "    (\"GTA\",\"V\"), (\"GTC\",\"V\"), (\"GTG\",\"V\"), (\"GTT\",\"V\"),\n",
    "    (\"GCA\",\"A\"), (\"GCC\",\"A\"), (\"GCG\",\"A\"), (\"GCT\",\"A\"),\n",
    "    (\"GAC\",\"D\"), (\"GAT\",\"D\"), (\"GAA\",\"E\"), (\"GAG\",\"E\"),\n",
    "    (\"GGA\",\"G\"), (\"GGC\",\"G\"), (\"GGG\",\"G\"), (\"GGT\",\"G\"),\n",
    "    (\"TCA\",\"S\"), (\"TCC\",\"S\"), (\"TCG\",\"S\"), (\"TCT\",\"S\"),\n",
    "    (\"TTC\",\"F\"), (\"TTT\",\"F\"), (\"TTA\",\"L\"), (\"TTG\",\"L\"),\n",
    "    (\"TAC\",\"Y\"), (\"TAT\",\"Y\"), (\"TAA\",\"stop\"), (\"TAG\",\"stop\"),\n",
    "    (\"TGC\",\"C\"), (\"TGT\",\"C\"), (\"TGA\",\"stop\"), (\"TGG\",\"W\")])\n",
    "\n",
    "function translate(dna_seq::String)\n",
    "    protein_seq = \"\"\n",
    "    dna_seq = uppercase(dna_seq)\n",
    "    \n",
    "    for i in 0:3:length(dna_seq)\n",
    "        if i+3 <= length(dna_seq) \n",
    "            if get(codon_dict, dna_seq[i+1:i+3], \"X\") != \"X\" && get(codon_dict, dna_seq[i+1:i+3], \"X\") != \"stop\"      \n",
    "                protein_seq = string(protein_seq, get(codon_dict, dna_seq[i+1:i+3], \"X\"))\n",
    "            end\n",
    "        end \n",
    "    end\n",
    "    return protein_seq\n",
    "end\n",
    "\n",
    "#println(\"Enter a DNA sequence\")\n",
    "#dna_seq = uppercase(readline())\n",
    "\n",
    "output = open(\"prot7.fsa\", \"w\")\n",
    "open(\"/Users/laurasansc/github/learning_julia/files/dna7.fsa\") do file\n",
    "    sequence = \"\"\n",
    "    protein_seq = \"\"\n",
    "    ID = \"\"\n",
    "    counter = 0\n",
    "    for line in eachline(file)\n",
    "        if occursin(\">\", line)\n",
    "            protein_seq = translate(sequence)    \n",
    "            write(output, \"\\n\",ID, \" Amino Acid Sequence\\n\")\n",
    "            for char in protein_seq\n",
    "                write(output, char)\n",
    "                counter += 1\n",
    "                if counter == 60\n",
    "                    write(output, \"\\n\")\n",
    "                    counter = 0\n",
    "                end\n",
    "                \n",
    "            end\n",
    "            counter = 0\n",
    "            ID = line\n",
    "        else\n",
    "            sequence = string(sequence,line)\n",
    "        end\n",
    "    end\n",
    " \n",
    "end\n",
    "close(output)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**3. You have made a program (let's call it the X-program), which as input takes a file of accession numbers, start10.dat and produces some output, which is in res10.dat. Now you count the lines in your input file and your output file and you discover that the line numbers do not match. Horror - your program does not produce output for some input. Now the assignment is to discover which accession numbers did not produce output. This can be done in various ways, but now you have to use a set. Print the results.**\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ex_7_3.jl <input> <output>\n",
    "\n",
    "\n",
    "input = ARGS[1]\n",
    "output = ARGS[2]\n",
    "\n",
    "infile = open(input, \"r\")\n",
    "outfile = open(output, \"r\")\n",
    "\n",
    "\n",
    "# input is 1 column, put into list\n",
    "start_ids = Set()\n",
    "start_ids = readlines(infile)\n",
    "\n",
    "\n",
    "# output is 5 columns, we want #2\n",
    "lines = readlines(outfile)\n",
    "res_ids = Set()\n",
    "\n",
    "for line in lines\n",
    "\tline = split(line, \"\\t\")\n",
    "\tpush!(res_ids, line[2])\n",
    "end\n",
    "\n",
    "# now compare sets\n",
    "\n",
    "difference = setdiff(start_ids, res_ids)\n",
    "# Check which are in the 1st file that are not in the 2nd\n",
    "first_check = intersect(difference, start_ids)\n",
    "\n",
    "println(first_check)\n",
    "println(difference)\n",
    "\n",
    "\n",
    "\n",
    "close(infile)\n",
    "close(outfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**4. In the file ex5.acc are a lot of accession numbers, where some are duplicates. Earlier we just removed the duplicates, now we should count them. Make a program that reads the file once, and prints a list (or writes a file) with the unique accession numbers and the number of occurrences in the file. A line should look like this: \"AC24677 2\", if this accession occurs twice in ex5.acc.**\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "acc_dict = Dict()\n",
    "# make a dictionary that counts the duplicates\n",
    "open(\"/Users/laurasansc/github/learning_julia/files/ex5.acc\") do file\n",
    "    for line in eachline(file)\n",
    "        if haskey(acc_dict, line) == false\n",
    "            acc_dict[line] = 1\n",
    "        else\n",
    "            acc_dict[line] += 1\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "outfile = open(\"out5.acc\", \"w\")\n",
    "for (key, value) in acc_dict\n",
    "    line = string(key, \"\\t\", value, \"\\n\")\n",
    "    write(outfile, line)\n",
    "end\n",
    "close(outfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**5. In the genbank files data?.gb you should extract the coding DNA sequence as you already did in 7.9. Next you have to display a list of codons USED in the coding sequence and the number of times they are used.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CODING SEQUENCE: 2 exons.\n",
      "DNA coding sequence:\n",
      "catggccctgtggatgcgcctcctgcccctgctggcgctgctggccctctggggacctgacccagccgcagcctttgtgaaccaacacctgtgcggctcacacctggtggaagctctctacctagtgtgcggggaacgaggcttcttctacacacccaagacccgccgggaggcagaggacctgcaggtggggcaggtggagctgggcgggggccctggtgcaggcagcctgcagcccttggccctggaggggtccctgcagaagcgtggcattgtggaacaatgctgtaccagcatctgctccctctaccagctggagaactactgcaacta \n",
      "\n",
      "\n",
      "TGT\t2\n",
      "CTT\t4\n",
      "AGT\t1\n",
      "CTA\t4\n",
      "GAC\t1\n",
      "GAA\t3\n",
      "TGC\t1\n",
      "ACC\t2\n",
      "ACA\t3\n",
      "CAA\t2\n",
      "GTG\t3\n",
      "CCT\t13\n",
      "GCT\t5\n",
      "GGC\t5\n",
      "AGG\t2\n",
      "CCC\t2\n",
      "AGC\t3\n",
      "CGG\t3\n",
      "GAT\t1\n",
      "GCC\t2\n",
      "GCG\t2\n",
      "TGA\t1\n",
      "GGT\t3\n",
      "CAG\t2\n",
      "GTC\t1\n",
      "CTC\t2\n",
      "TCT\t1\n",
      "CTG\t4\n",
      "CGC\t1\n",
      "ATG\t1\n",
      "GGA\t8\n",
      "ACG\t1\n",
      "TAC\t1\n",
      "CCA\t2\n",
      "CAT\t3\n",
      "AGA\t1\n",
      "TGG\t2\n",
      "GGG\t5\n",
      "CAC\t1\n",
      "CCG\t2\n",
      "GCA\t4\n"
     ]
    }
   ],
   "source": [
    "infile = \"/Users/laurasansc/github/learning_julia/files/data1.gb\"\n",
    "\n",
    "open(infile) do file\n",
    "    \n",
    "    # define variables\n",
    "    dna_seq = \"\"\n",
    "    cds_exons = []\n",
    "    cds_exons_initial = []\n",
    "    flag = false\n",
    "    \n",
    "    for ln in eachline(file)\n",
    "        if occursin(\"CDS\", ln)\n",
    "            cds_exons_search = match(r\"join\\(([\\d\\.,]+)\\)\", ln)\n",
    "            cds_exons_initial = split(cds_exons_search.captures[1], \",\")\n",
    "            for item in cds_exons_initial\n",
    "                push!(cds_exons, split(item, \"..\"))\n",
    "            end\n",
    "        # get all DNA\n",
    "        elseif occursin(\"//\", ln)\n",
    "            flag = false\n",
    "        elseif flag == true\n",
    "            dna_search = match(r\"\\b\\d+\\b([\\D\\s]+)\\b\", ln)\n",
    "            dna_seq = string(dna_seq, dna_search.captures[1])\n",
    "        elseif occursin(\"ORIGIN\", ln)\n",
    "            flag = true\n",
    "        end\n",
    "    end\n",
    "\n",
    "    dna_seq = replace(dna_seq, \" \" => \"\")\n",
    "    cds_seq = \"\"\n",
    "    \n",
    "    for item in cds_exons\n",
    "        start_pos = parse(Int64, item[1]) - 1\n",
    "        final_pos = parse(Int64, item[2]) - 1\n",
    "        cds_seq = string(cds_seq, dna_seq[start_pos:final_pos])\n",
    "    end\n",
    "    codons = Dict()\n",
    "    for j in 1:3:length(cds_seq)\n",
    "        if j < length(cds_seq) && (j+3) < length(cds_seq)\n",
    "            if haskey(codons, uppercase(cds_seq[j:j+2])) == false\n",
    "                if length(cds_seq[j:j+2]) == 3\n",
    "                    codons[uppercase(cds_seq[j:j+2])] = 1    \n",
    "                end\n",
    "            else\n",
    "                codons[uppercase(cds_seq[j:j+2])] += 1\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    # stdout the final output#\n",
    "    num_exons = length(cds_exons)\n",
    "    println(\"CODING SEQUENCE: $num_exons exons.\\nDNA coding sequence:\\n$cds_seq \\n\\n\")\n",
    "    \n",
    "    for (key, value) in codons\n",
    "        println(key, \"\\t\", value)\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**6. In the data1.gb file there are 6 references (to articles). Make a program that extracts all authors from the references, eliminates those that are duplicates and print the list of authors. You will notice that some authors seems to be the same person using different initials. You should only consider a person a duplicate if the name matches excatly. This should also work for the other Genbank entries. Beware: there traps in this exercise, check your output properly. You are free to use sets/dictionaries or not in this exercise.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# julia ex_7_6.jl <input file>\n",
    "\n",
    "input = ARGS[1]\n",
    "\n",
    "# Initialize\n",
    "\n",
    "open(input) do file\n",
    "    authors = []\n",
    "    flag = false\n",
    "    for ln in eachline(file)\n",
    "        if occursin(\"TITLE\", ln)\n",
    "            flag = false\n",
    "        end\n",
    "        if flag == true\n",
    "            push!(authors, ln)\n",
    "        end\n",
    "        if occursin(\"AUTHORS\", ln)\n",
    "            flag = true\n",
    "            authors_search = match(r\"AUTHORS\\s+(.+)\", ln)\n",
    "            push!(authors, authors_search.captures[1]) \n",
    "        end\n",
    "    end\n",
    "\n",
    "    authors_clean = []\n",
    "    for item in authors\n",
    "        #item = replace(item, \" and \" => \", \")\n",
    "        item = replace(item, \" and\" => \", \")\n",
    "        item = strip(item)\n",
    "        #println(item)\n",
    "        sub_list = split(item, \", \")\n",
    "        for sub_item in sub_list\n",
    "            sub_item = strip(sub_item)\n",
    "            push!(authors_clean, sub_item)\n",
    "        end\n",
    "\n",
    "    end\n",
    "    println(unique(authors_clean))\n",
    "end"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.5.0",
   "language": "julia",
   "name": "julia-1.5"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.5.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
